src/util/tendril.rs

```diff
@@ -0,0 +1,922 @@
+    /// either all match or all don't match the predicate,
+    /// and also return whether they match.
+    ///
+    /// Returns `None` on an empty string.
+    pub fn char_run<Pred>(&self, mut pred: Pred) -> Option<(u32, bool)>
+        where Pred: FnMut(char) -> bool,
+    {
+        let (first, rest) = unwrap_or_return!(self.slice_shift_char(), None);
+        let matches = pred(first);
+
+        for (idx, ch) in rest.char_indices() {
+            if matches != pred(ch) {
+                return Some(((idx + first.len_utf8()) as u32, matches));
+            }
+        }
+        Some((self.len32(), matches))
+    }
+
+    /// Promotes the tendril to owning its content, and get a
+    /// mutable reference.
+    ///
+    /// This is unsafe because the user must not exceed the 1 GB
+    /// size limit!
+    #[inline]
+    unsafe fn to_mut<'a>(&'a mut self) -> &'a mut String {
+        match self.0 {
+            Owned(ref mut s) => return s,
+            _ => (),
+        }
+
+        self.0 = Owned(String::from_str(self));
+        match self.0 {
+            Owned(ref mut s) => s,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    fn check_len(&self) {
+        if self.len() > TENDRIL_MAX_LEN as usize {
+            panic!("tendril exceeded 1 GB");
+        }
+    }
+
+    /// Push a character onto the end of the tendril.
+    #[inline]
+    pub fn push(&mut self, c: char) {
+        if self.is_empty() {
+            if let Some(a) = c.to_ascii_opt() {
+                self.0 = Ascii(a.to_u8());
+                return;
+            }
+        }
+        unsafe {
+            self.to_mut().push(c);
+        }
+        self.check_len();
+    }
+
+    /// Push a string onto the end of the tendril.
+    #[inline]
+    pub fn push_str(&mut self, rhs: &str) {
+        match rhs.len() {
+            0 => return,
+            1 if self.is_empty() => {
+                if let Some(n) = as_single_ascii(rhs) {
+                    self.0 = Ascii(n);
+                    return;
+                }
+            }
+            n if n > TENDRIL_MAX_LEN as usize => {
+                panic!("attempted to extend tendril by more than 1 GB");
+            }
+
+            // Otherwise, 2 * TENDRIL_MAX_LEN does not overflow u32.
+            _ => (),
+        }
+        unsafe {
+            self.to_mut().push_str(rhs);
+        }
+        self.check_len();
+    }
+
+    /// Push another tendril onto the end of the tendril.
+    #[inline]
+    pub fn push_tendril(&mut self, rhs: Tendril) {
+        if rhs.is_empty() {
+            return;
+        }
+
+        if self.is_empty() {
+            *self = rhs;
+            return;
+        }
+
+        // Try to merge adjacent Iobufs.
+        if let (&mut Tendril(Shared(ref mut a)), &Tendril(Shared(ref b)))
+            = (&mut *self, &rhs)
+        {
+            if a.extend_with(b).is_ok() {
+                return;
+            }
+        }
+
```

Comment:
I think everything below this point should be out in a function marked `#[cold]`.